sdk/src/Core/Amazon.Runtime/Credentials/FallbackCredentialsFactory.cs

```diff
@@ -124,61 +136,84 @@ public static AWSCredentials GetCredentials()
 
         public static AWSCredentials GetCredentials(bool fallbackToAnonymous)
         {
-            if (cachedCredentials != null)
-                return cachedCredentials;
-
-            List<Exception> errors = new List<Exception>();
-
-            foreach (CredentialsGenerator generator in CredentialsGenerators)
+            try
             {
-                try
+                cachedCredentialsLock.EnterReadLock();
+                if (cachedCredentials != null)
                 {
-                    cachedCredentials = generator();
+                    return cachedCredentials;
                 }
-                // Breaking the FallbackCredentialFactory chain in case a ProcessAWSCredentialException exception 
-                // is encountered. ProcessAWSCredentialException is thrown by the ProcessAWSCredential provider
-                // when an exception is encountered when running a user provided process to obtain Basic/Session 
-                // credentials. The motivation behind this is that, if the user has provided a process to be run
-                // he expects to use the credentials obtained by running the process. Therefore the exception is
-                // surfaced to the user.
-                catch (ProcessAWSCredentialException)
+            }
+            finally
+            {
+                cachedCredentialsLock.ExitReadLock();
+            }
+            
+            try
```

Comment:
I don't think the topmost `try` block here is needed but I'll let @normj chime in. When norm said "Here is an another example in the SDK doing the lock double check" I think he meant something like:

```
if(cachedCredentials != null)
{
  return cachedCredentials
}
List<Exception> errors = new List<Exception>();
  try{
    cachedCredentials.EnterWriteLock();
    if (cachedCredentials != null)
    {
        return cachedCredentials;
    }
     foreach (CredentialsGenerator generator in CredentialsGenerators)
     {
       try
       {
          cachedCredentials = generator();
        }
        .... skipping logic here     
     }
  }
  finally
   {
     cachedCredentialsLock.ExitWriteLock();
   }
return cachedCredentials;

```

Reply:
This would've worked if we never set the cached credentials to null, both during the loop and via the reset method. Even then, there are JIT optimizations that might cause issues, hence the suggestion to use of a readwritelockslim in the first place.

Reply:
Okay, in that case I don't see any issues. Approving